ksql-serde/src/main/java/io/confluent/ksql/serde/json/KsqlJsonSerializer.java

```diff
@@ -47,31 +57,160 @@ public void configure(final Map<String, ?> props, final boolean isKey) {
     if (data == null) {
       return null;
     }
-
     try {
-      return objectMapper.writeValueAsBytes(dataToMap(data));
+      Struct struct = new Struct(schema);
+      for (int i = 0; i < data.getColumns().size(); i++) {
+        if (schema.fields().get(i).schema().type() == Schema.Type.STRUCT) {
+          struct.put(schema.fields().get(i), updateStructSchema(
+              (Struct) data.getColumns().get(i),
+              schema.fields().get(i).schema()));
+        } else {
+          struct.put(schema.fields().get(i), data.getColumns().get(i));
+        }
+
+      }
+
+      byte[] serialized = jsonConverter.fromConnectData(topic, schema, struct);
+      return serialized;
     } catch (Exception e) {
       throw new SerializationException("Error serializing JSON message", e);
     }
   }
 
-  private Map<String, Object> dataToMap(final GenericRow data) {
-    if (data == null) {
-      return null;
+  private Schema makeOptional(Schema schema) {
+    switch (schema.type()) {
+      case BOOLEAN:
+        return Schema.OPTIONAL_BOOLEAN_SCHEMA;
+      case INT32:
+        return Schema.OPTIONAL_INT32_SCHEMA;
+      case INT64:
+        return Schema.OPTIONAL_INT64_SCHEMA;
+      case FLOAT64:
+        return Schema.OPTIONAL_FLOAT64_SCHEMA;
+      case STRING:
+        return Schema.OPTIONAL_STRING_SCHEMA;
+      case ARRAY:
+        return SchemaBuilder.array(makeOptional(schema.valueSchema())).optional().build();
+      case MAP:
+        return SchemaBuilder.map(schema.keySchema(), makeOptional(schema.valueSchema()))
+            .optional().build();
+      case STRUCT:
+        SchemaBuilder schemaBuilder = new SchemaBuilder(Schema.Type.STRUCT);
+        schema.fields()
+            .forEach(field -> schemaBuilder.field(field.name(), makeOptional(field.schema())));
+        return schemaBuilder.optional().build();
+      default:
+        throw new KsqlException("Invalid type in schema: " + schema);
+    }
+  }
+
+  /**
+   * Temporary work around until schema comparison is fixed.
```

Comment:
I think we should try and understand and document exactly why this is necessary. Is it the optional flag that doesn't match? If so we should try and set optional correctly in the engine itself. I've done this for structs in the avro serde pr.

Reply:
Yes, this is a bit tricky since Connect enforces optional requirement. I'm making changes in the code so all of our columns are optional. This will make sure we can have null columns. We may change it when we add `NOT NULL` to the stream/table declaration.